File: python/pex/wx/shortcut.py
from typing import (
    Any,
    Callable,
    Iterable,
    Union,
    Tuple,
    List,
    Dict)

import itertools

import wx
import attr
from typing_extensions import Protocol
from .. import pex


keyStringByWxDefine = {
    wx.WXK_DELETE: 'DELETE',
    wx.WXK_BACK: 'BACK',
    wx.WXK_INSERT: 'INSERT',
    wx.WXK_RETURN: 'RETURN',
    wx.WXK_PAGEUP: 'PGUP',
    wx.WXK_PAGEDOWN: 'PGDN',
    wx.WXK_LEFT: 'LEFT',
    wx.WXK_RIGHT: 'RIGHT',
    wx.WXK_UP: 'UP',
    wx.WXK_DOWN: 'DOWN',
    wx.WXK_HOME: 'HOME',
    wx.WXK_END: 'END',
    wx.WXK_SPACE: 'SPACE',
    wx.WXK_TAB: 'TAB',
    wx.WXK_ESCAPE: 'ESCAPE'}


modifierStringByWxAccel = {
    wx.ACCEL_NORMAL: '',
    wx.ACCEL_SHIFT: 'SHIFT',
    wx.ACCEL_CMD: 'CTRL',
    wx.ACCEL_ALT: 'ALT'}


modifierOrder = (wx.ACCEL_CMD, wx.ACCEL_SHIFT, wx.ACCEL_ALT)


class Key:
    asciiCode_: int

    def __init__(self, character: Union[int, str]) -> None:
        if isinstance(character, int):
            self.asciiCode_ = character
        else:
            self.asciiCode_ = ord(character)

    def __repr__(self) -> str:
        return keyStringByWxDefine.get(
            self.asciiCode_,
            chr(self.asciiCode_))

    def __int__(self) -> int:
        return self.asciiCode_


def GetModifierString(modifierBitfield: int) -> str:
    """
    Returns a string describing all modifiers in the bitfield.

    modifierBitfield may be bitwise combination of modifiers. Check the
    modifierBitfield against each accel define to build the description string.
    """
    return '+'.join([
        modifierStringByWxAccel[modifier]
        for modifier in modifierOrder
        if modifierBitfield & modifier])


@attr.s(auto_attribs=True, eq=False, slots=True)
class Shortcut:
    signal: pex.InterfaceSignal
    modifier: int
    ascii: str
    description: str
    longDescription: str

    @property
    def name(self) -> str:
        return self.signal.name


class ShortcutMethods:
    signal_: pex.InterfaceSignal
    id_: int
    modifier_: int
    key_: Key
    description_: str
    longDescription_: str

    def __init__(self, shortcut: Shortcut) -> None:
        self.signal_ = shortcut.signal
        self.id_ = wx.Window.NewControlId()
        self.modifier_ = shortcut.modifier
        self.key_ = Key(shortcut.ascii)
        self.description_ = shortcut.description
        self.longDescription_ = shortcut.longDescription

    def AddToMenu(self, menu: wx.Menu) -> None:
        menu.Append(
            wx.MenuItem(
                menu,
                self.id_,
                self.GetMenuItemLabel(),
                self.longDescription_))

    def GetMenuItemLabel(self) -> str:
        modifier = GetModifierString(self.modifier_)

        if len(modifier) > 0:
            return u'{}\t{}+{}'.format(self.description_, modifier, self.key_)
        else:
            return u'{}\t{}'.format(self.description_, self.key_)

    def GetAcceleratorTableEntry(self) -> Tuple[int, int, int]:
        return (self.modifier_, int(self.key_), self.id_)

    def OnEventMenu(self, ignored: wx.CommandEvent) -> None:
        self.signal_.Trigger()

    def GetId(self) -> int:
        return self.id_


class HasWxShortcutMethods(Protocol):
    acceleratorShortcutMethods_: List[ShortcutMethods]
    menuShortcutMethodsByName_: Dict[str, ShortcutMethods]
    acceleratorTable_: wx.AcceleratorTable

    def SetAcceleratorTable(
            self,
            acceleratorTable_: wx.AcceleratorTable) -> None:
        ...

    def Bind(
            self,
            binder: Any,
            callback: Callable[[wx.CommandEvent], None],
            id: int = wx.ID_ANY) -> None: # pylint: disable=redefined-builtin
        ...


class ShortcutMixin:
    def __init__(
            self: HasWxShortcutMethods,
            acceleratorShortcuts: Iterable[Shortcut],
            menuShortcuts: Iterable[Shortcut]):

        self.acceleratorShortcutMethods_ = [
            ShortcutMethods(shortcut) for shortcut in acceleratorShortcuts]

        self.menuShortcutMethodsByName_ = {
            shortcut.name: ShortcutMethods(shortcut)
            for shortcut in menuShortcuts}

        self.acceleratorTable_ = wx.AcceleratorTable([
            shortcutMethods.GetAcceleratorTableEntry()
            for shortcutMethods in self.acceleratorShortcutMethods_])

        self.SetAcceleratorTable(self.acceleratorTable_)

        for shortcut in itertools.chain(
                self.acceleratorShortcutMethods_,
                self.menuShortcutMethodsByName_.values()):

            self.Bind(
                wx.EVT_MENU,
                shortcut.OnEventMenu,
                id=shortcut.GetId())

    def AddShortcutToMenu(self, shortcutName: str, menu: wx.Menu) -> None:
        self.menuShortcutMethodsByName_[shortcutName].AddToMenu(menu)

File: pex/wx/shortcut.cpp
#include "pex/wx/shortcut.h"


namespace pex
{


namespace wx
{


std::string Key::GetString() const
{
    if (this->keyStringByKeyCode_.count(this->code_))
    {
        return this->keyStringByKeyCode_.at(this->code_);
    }
    
    if (this->code_ > 127)
    {
        throw std::runtime_error("Unsupported key code.");
    }

    return std::string(1, static_cast<char>(this->code_));
}

int Key::GetInt() const { return this->code_; }

const std::map<int, std::string> Key::keyStringByKeyCode_
{
    {WXK_DELETE, "DELETE"},
    {WXK_BACK, "BACK"},
    {WXK_INSERT, "INSERT"},
    {WXK_RETURN, "RETURN"},
    {WXK_PAGEUP, "PGUP"},
    {WXK_PAGEDOWN, "PGDN"},
    {WXK_LEFT, "LEFT"},
    {WXK_RIGHT, "RIGHT"},
    {WXK_UP, "UP"},
    {WXK_DOWN, "DOWN"},
    {WXK_HOME, "HOME"},
    {WXK_END, "END"},
    {WXK_SPACE, "SPACE"},
    {WXK_TAB, "TAB"},
    {WXK_ESCAPE, "ESCAPE"}
};


std::string GetModifierString(int modifierBitfield)
{
    static const std::map<wxAcceleratorEntryFlags, std::string>
    modifierStringByWxAccel
    {
        {wxACCEL_NORMAL, ""},
        {wxACCEL_SHIFT, "SHIFT"},
        {wxACCEL_CTRL, "CTRL"},
        {wxACCEL_ALT, "ALT"},
        {wxACCEL_RAW_CTRL, "RAWCTRL"}
    };

    static const std::vector<wxAcceleratorEntryFlags>

    modifierOrder
    {
        wxACCEL_CTRL,
        wxACCEL_SHIFT,
        wxACCEL_ALT
#ifdef __APPLE__
        , wxACCEL_RAW_CTRL
#endif 
    };

    std::vector<std::string> modifiers;

    for (auto modifier: modifierOrder)
    {
        if (modifierBitfield & modifier)
        {
            modifiers.push_back(modifierStringByWxAccel.at(modifier));
        }
    }

    return jive::strings::Join(modifiers.begin(), modifiers.end(), '+');
}


void Shortcut::AddToMenu(wxMenu *menu)
{
    menu->Append(
        this->menuItem_ = new wxMenuItem(
            menu,
            this->id_,
            this->GetMenuItemLabel_(),
            this->longDescription_));
}

wxAcceleratorEntry Shortcut::GetAcceleratorEntry() const
{
    return wxAcceleratorEntry(
        this->modifier_,
        this->key_.GetInt(),
        this->id_,
        this->menuItem_);
}

int Shortcut::GetModifier() const { return this->modifier_; }

int Shortcut::GetKeyAsInt() const { return this->key_.GetInt(); }

int Shortcut::GetId() const { return this->id_; }

void Shortcut::OnEventMenu()
{
    this->signal_.Trigger();
}

wxString Shortcut::GetMenuItemLabel_() const
{
    auto modifier = GetModifierString(this->modifier_);
    
    std::string result(this->description_);

    if (!modifier.empty())
    {
        jive::strings::Concatenate(
            &result,
            '\t',
            modifier,
            '+',
            this->key_.GetString());
    }
    else
    {
        jive::strings::Concatenate(
            &result,
            '\t',
            this->key_.GetString());
    }
    
    return wxString(result);
}


ShortcutsBase::ShortcutsBase(
    wxWindow *window,
    const ShortcutsByMenu &shortcutsByMenu)
    :
    window_(window),
    shortcutsByMenu_(shortcutsByMenu)
{

}

ShortcutsBase::~ShortcutsBase()
{
    PEX_LOG("Unbind shortcuts");

    for (auto & [menuName, shortcuts]: this->shortcutsByMenu_)
    {
        this->UnbindShortcuts(shortcuts);
    }
}

void ShortcutsBase::BindShortcuts(Shortcuts &shortcuts)
{
    for (auto &shortcut: shortcuts)
    {
        this->window_->Bind(
            wxEVT_MENU,
            ShortcutFunctor(shortcut),
            shortcut.GetId());
    }
}

void ShortcutsBase::UnbindShortcuts(Shortcuts &shortcuts)
{
    for (auto &shortcut: shortcuts)
    {
        this->window_->Unbind(
            wxEVT_MENU,
            ShortcutFunctor(shortcut),
            shortcut.GetId());
    }
}


MenuShortcuts::MenuShortcuts(
    wxWindow *window,
    ShortcutsByMenu &shortcutsByMenu)
    :
    ShortcutsBase(window, shortcutsByMenu),
    menuBar_(std::make_unique<wxMenuBar>())
{
    for (auto & [menuName, shortcuts]: this->shortcutsByMenu_)
    {
        auto menu = std::make_unique<wxMenu>();
        this->AddToMenu(menu.get(), shortcuts);

        this->menuBar_->Append(
            menu.release(),
            wxString(std::forward<decltype(menuName)>(menuName)));

        this->BindShortcuts(shortcuts);
    }
}

wxMenuBar *MenuShortcuts::GetMenuBar()
{
    return this->menuBar_.release();
}

void MenuShortcuts::AddToMenu(
    wxMenu *menu,
    Shortcuts &shortcuts)
{
    for (auto &shortcut: shortcuts)
    {
        shortcut.AddToMenu(menu);
    }
}


AcceleratorShortcuts::AcceleratorShortcuts(
    wxWindow *window,
    const ShortcutsByMenu &shortcutsByMenu)
    :
    ShortcutsBase(window, shortcutsByMenu),
    acceleratorTable_()
{
    std::vector<std::vector<wxAcceleratorEntry>> entries;
    size_t entryCount = 0;

    for (auto & [menuName, shortcuts]: this->shortcutsByMenu_)
    {
        auto shortcutEntries = CreateAcceleratorEntries(shortcuts); 
        entryCount += shortcutEntries.size();
        entries.push_back(shortcutEntries);
        this->BindShortcuts(shortcuts);
    }

    std::vector<wxAcceleratorEntry> allEntries;
    allEntries.reserve(entryCount);

    for (auto &it: entries)
    {
        allEntries.insert(allEntries.end(), it.begin(), it.end());
    }

    assert(allEntries.size() == entryCount);

    this->acceleratorTable_ = wxAcceleratorTable(
        static_cast<int>(entryCount),
        allEntries.data());
}

const wxAcceleratorTable & AcceleratorShortcuts::GetAcceleratorTable() const
{
    return this->acceleratorTable_;
}

std::vector<wxAcceleratorEntry>
AcceleratorShortcuts::CreateAcceleratorEntries(const Shortcuts &shortcuts)
{
    std::vector<wxAcceleratorEntry> entries;
    entries.reserve(shortcuts.size());

    for (auto &shortcut: shortcuts)
    {
        entries.push_back(shortcut.GetAcceleratorEntry());
    }

    return entries;
}


} // end namespace wx


} // end namespace pex
